### PR TITLE
[SPARK-21263][SQL] Do not allow partially parsing double and floats via NumberFormat in CSV

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR proposes to remove `NumberFormat.parse` use to disallow a case of partially parsed data. For example,

```
scala> spark.read.schema("a DOUBLE").option("mode", "FAILFAST").csv(Seq("10u12").toDS).show()
+----+
|   a|
+----+
|10.0|
+----+
```

## How was this patch tested?

Unit tests added in `UnivocityParserSuite` and `CSVSuite`.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @srowen and @falaki, could you take a look and see if I understood correctly?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala [hunk 1]
```diff
@@ -111,19 +111,15 @@ class UnivocityParser(
         case options.nanValue => Float.NaN
         case options.negativeInf => Float.NegativeInfinity
         case options.positiveInf => Float.PositiveInfinity
-        case datum =>
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: BTW, it looks we are not using `NumberFormat.parse` in schema inference - https://github.com/apache/spark/blob/7e5359be5ca038fdb579712b18e7f226d705c276/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala#L141

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 2]
```diff
@@ -1174,4 +1174,25 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
         }
       }
   }
+
+  test("Do not partially lose data when parsing float and double") {
```
Comment (falaki) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: I suggest a better description for this test and please include the JIRA number. E.g., 
```SPARK-21263: Invalid float and double are handled correctly in different modes```

**Review by falaki — CHANGES_REQUESTED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala (hunk 2 above)
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: Sure, thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParserSuite.scala [hunk 3]
```diff
@@ -130,17 +130,17 @@ class UnivocityParserSuite extends SparkFunSuite {
...
-      assert(parser.makeConverter("_1", DoubleType, options = options).apply("1,00") == 100.0)
-    } finally {
-      Locale.setDefault(originalLocale)
-    }
+  test("Throws exception for casting an invalid string to Float and Double Types") {
+    val options = new CSVOptions(Map.empty[String, String], "GMT")
+    var message = intercept[NumberFormatException] {
+      parser.makeConverter("_1", FloatType, options = options).apply("10u000")
+    }.getMessage
+    assert(message.contains("10u000"))
```
Comment (falaki) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParserSuite.scala]: Is there some more specific error we could check for?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParserSuite.scala [hunk 4]
```diff
@@ -130,17 +130,17 @@ class UnivocityParserSuite extends SparkFunSuite {
...
+
+    message = intercept[NumberFormatException] {
+      parser.makeConverter("_1", DoubleType, options = options).apply("10u000")
+    }.getMessage
+    assert(message.contains("10u000"))
```
Comment (falaki) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParserSuite.scala]: dito

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 5]
```diff
@@ -1174,4 +1174,25 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
...
   }
+
+  test("SPARK-21263: Invalid float and double are handled correctly in different modes") {
+    val exception = intercept[SparkException] {
+      spark.read.schema("a DOUBLE")
+        .option("mode", "FAILFAST")
+        .csv(Seq("10u12").toDS())
+        .collect()
+    }
+    assert(exception.getMessage.contains("10u12"))
```
Comment (falaki) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: Can we check for more specific error message?

**Review by falaki — COMMENTED** [review]
Minor suggestions. Otherwise LGTM

**Comment by HyukjinKwon** [comment]
Thank you @falaki. I just updated.

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**